credential/cargo-credential/src/error.rs

```diff
@@ -0,0 +1,193 @@
+use serde::{Deserialize, Serialize};
+use std::error::Error as StdError;
+use thiserror::Error as ThisError;
+
+/// Credential provider error type.
+/// 
+/// `UrlNotSupported` and `NotFound` errors both cause Cargo
+/// to attempt another provider, if one is available. The other 
+/// variants are fatal.
+/// 
+/// Note: Do not add a tuple variant, as it cannot be serialized.
+#[derive(Serialize, Deserialize, ThisError, Debug)]
+#[serde(rename_all = "kebab-case", tag = "kind")]
+#[non_exhaustive]
+pub enum Error {
+    #[error("registry not supported")]
+    UrlNotSupported,
+    #[error("credential not found")]
+    NotFound,
+    #[error("requested operation not supported")]
+    OperationNotSupported,
+    #[error("protocol version {version} not supported")]
+    ProtocolNotSupported { version: u32 },
+    #[error(transparent)]
+    #[serde(with = "error_serialize")]
+    Other(Box<dyn StdError + Sync + Send>),
```

Comment:
And whether messages and programmatic variants are actually mutually exclusive or not

Reply:
Oh, most of this moved so I guess this is out of scope for this PR though something for us to keep in mind

Reply:
I wanted to have variants where Cargo may want to do something differently based on the provider output. Everything else should go into the `Other` variant.

`UrlNotSupported` and `NotFound` both cause Cargo to try the next provider in the global credential provider list rather than erroring out.

I did reduce the number of variants in this change, but we could possibly reduce it further. I kept `ProtocolNotSupported` and `OperationNotSupported` in case we wanted to handle those differently in the future.

Reply:
Makes sense.  The fact that we only document behavior differences with some variants is partly why I asked.